### PR TITLE
MAINT: Unify reductions in fromnumeric.py

### LINKED ISSUE #10652: DOC/BUG: np.sum supports generators, unlike the other reduction helpers
```python
>>> np.sum(a for a in range(10))
45
>>> np.prod(a for a in range(10))
<generator object <genexpr> at 0x0000023BAF7ED1A8>
```

Generators are not considered `array_like`, so the result for `sum` does not match the documentation.

Introduced in a13aad3ac33b629f3e696b4d4d5dbf4b5605d567

### PR DESCRIPTION
**Pull request by hameerabbasi**
Inspired while fixing #10652.

### CODE REVIEW

File: numpy/core/fromnumeric.py
Comment (eric-wieser) [inline numpy/core/fromnumeric.py]: In my opinion we want to deprecate this, not add support for it to the other function:

* It's not documented
* It completely ignores the `axis`, `dtype`, and `keepdims` arguments
* There's no precedent for accepting generators elsewhere in numpy

File: numpy/core/fromnumeric.py
Comment (eric-wieser) [inline numpy/core/fromnumeric.py]: I actually think we want to deprecate this too - do we need `__getattr__`-style duck-typing now that we have `__array_ufunc__`?

**Comment by eric-wieser** [comment]
I do like the cleanup, but I don't want to enable the spread of questionable behaviour either

File: numpy/core/fromnumeric.py
Comment (hameerabbasi) [reply numpy/core/fromnumeric.py]: If a library wants to support both old and new versions of Numpy, it's needed.

File: numpy/core/fromnumeric.py
Comment (eric-wieser) [inline numpy/core/fromnumeric.py]: For consistency either this should be `um.multiply` or the other ones should be `np.minimum` etc. I'd be inclined to go with the `np.` form for clarity

File: numpy/core/fromnumeric.py
Comment (hameerabbasi) [reply numpy/core/fromnumeric.py]: I'll remove it completely. Hopefully no tests fail.

File: numpy/core/fromnumeric.py
Comment (mhvk) [inline numpy/core/fromnumeric.py]: I would just stick with `keepdims` at the top and keep the test as is. This is only meant as a work-around for old code that does not support `keepdims`; if we add other keywords, we can just expand.

Comment (hameerabbasi) [reply numpy/core/fromnumeric.py]: I just thought this makes it easy to add kwargs to reductions, like `initializer` in my other PR.

File: numpy/core/fromnumeric.py
Comment (eric-wieser) [inline numpy/core/fromnumeric.py]: You should leave this code here in `sum` and add:
```python
# Numpy 1.15, 2018-02-25
warnings.warn(
    "Calling np.sum(generator) is deprecated, and in the future will give a different result. "
    "Use np.sum(np.from_iter(generator)) or the python sum builtin instead."
    FutureWarning, stacklevel=2)
```

Comment (hameerabbasi) [reply numpy/core/fromnumeric.py]: Since it's undocumented, it should be okay to remove it entirely, no? I doubt much people rely on it.

**Comment by hameerabbasi** [comment]
A few comments.

File: numpy/core/fromnumeric.py
Comment (eric-wieser) [reply numpy/core/fromnumeric.py]: Unfortunately that's rarely the case. In this case, it's very likely that someone's old code is doing
```
from numpy import *

def use_what_I_thought_was_the_builtin_sum():
    return sum(i for i in range(10))
```
We can't just break their code without warning, much though it's tempting to.

File: numpy/core/fromnumeric.py
Comment (mhvk) [reply numpy/core/fromnumeric.py]: I hadn't seen that PR when I reviewed...

File: numpy/core/fromnumeric.py
Comment (hameerabbasi) [reply numpy/core/fromnumeric.py]: Done!

File: numpy/core/fromnumeric.py
Comment (hameerabbasi) [reply numpy/core/fromnumeric.py]: You're right. Done!

File: numpy/core/fromnumeric.py
Comment (hameerabbasi) [reply numpy/core/fromnumeric.py]: Done!

File: numpy/core/fromnumeric.py
Comment (eric-wieser) [inline numpy/core/fromnumeric.py]: This moved because your IDE alphabetized it, I'm guessing? Just curious

File: numpy/core/fromnumeric.py
Comment (eric-wieser) [inline numpy/core/fromnumeric.py]: This could do with a comment justifying it - especially since it wasn't needed previously

File: numpy/core/fromnumeric.py
Comment (eric-wieser) [reply numpy/core/fromnumeric.py]: Actually, is this import needed at all now?

Comment (hameerabbasi) [reply numpy/core/fromnumeric.py]: Yes, for mean, etc.

Comment (hameerabbasi) [reply numpy/core/fromnumeric.py]: And no, I actually removed it and then had to put it back. I can put it back where it was if it's an issue.

File: numpy/core/fromnumeric.py
Comment (eric-wieser) [inline numpy/core/fromnumeric.py]: This comment normally goes immediately above the warning.

You should add a test for this deprecation in `core/tests/test_deprecations.py` too, along with the same comment with date and version

File: numpy/core/fromnumeric.py
Comment (eric-wieser) [inline numpy/core/fromnumeric.py]: I think this line should stay

File: numpy/core/fromnumeric.py
Comment (eric-wieser) [reply numpy/core/fromnumeric.py]: It's slightly better to minimize noise in the diffs, but it's not worth your time to make a new commit to fix that

File: numpy/core/tests/test_deprecations.py
Comment (eric-wieser) [inline numpy/core/tests/test_deprecations.py]: Would be better not to commit these random style changes in the same PR as a feature.

If you're not aware, try using `git add -p` which lets you choose which changes you actually want to commit

Comment (hameerabbasi) [reply numpy/core/tests/test_deprecations.py]: Thanks for the tip! Should I reverse these (it is significant).

Comment (hameerabbasi) [reply numpy/core/tests/test_deprecations.py]: Most of these come from PEP8, very few are PyCharm specific.

Comment (eric-wieser) [reply numpy/core/tests/test_deprecations.py]: Random as in unrelated to the change, not random stylistically. Keeping diffs small makes blames easier in future (ie, "which commit introduced this deprecation" is harder to answer if you touch a lot of lines). So yes, reversing these would be good.

Comment (hameerabbasi) [reply numpy/core/tests/test_deprecations.py]: Done!

File: numpy/core/fromnumeric.py
Comment (hameerabbasi) [reply numpy/core/fromnumeric.py]: Should we document the deprecation? I shouldn't think it's necessary, since the functionality itself wasn't documented in the first place.

File: numpy/core/fromnumeric.py
Comment (eric-wieser) [inline numpy/core/fromnumeric.py]: `method` should be named `ufunc`, and `name` should be named `method`, for consistency with `_wrapfunc`

**Comment by hameerabbasi** [comment]
I notice that reductions use the attribute by default, instead of the `__array_ufunc__` with `method=reduce`. Would it be wise to change this?

**Comment by eric-wieser** [comment]
If they used the attribute before, let's have them continue to use it for now.

The deprecation needs a release note.

**Comment by hameerabbasi** [comment]
I actually meant we should support both, but default to the ufunc.

**Comment by eric-wieser** [comment]
Perhaps, but not in this PR. Let's not let the scope creep too much. I already think we might want to split the deprecation into a separate commit, but I can handle that when I merge

**Comment by hameerabbasi** [comment]
Something like: 
```python
try:
    ufunc.reduce(...)
except e:
    Process attribute case here
```

This will be helpful in the future as well, for duck arrays.

**Comment by eric-wieser** [comment]
@hameerabbasi: I'm gonna rewrite the history now, actually, and get the deprecation commit in a separate PR - it'll still have your name on it, but I think we should review that separately.

**Comment by eric-wieser** [comment]
Alright, deprecation is gone

**Review by eric-wieser — APPROVED** [review]
Substantial cleanup, thanks @hameerabbasi. Will merge in the next few days unless someone else has something to say

**Comment by hameerabbasi** [comment]
All good with me. As a side note, how DO you decide during a merge who owns the commit?

**Comment by eric-wieser** [comment]
I assume by that, you're asking how that's handled by git? Commits have two "owner"s - the "committer", and the "author". The two avatars in the commit above show you as the author (which you are), but me as the commiter (because I rebased your commit). Both can be set to whatever you want in principle.

**Review by mhvk — APPROVED** [review]
Looks all great to me. Thanks, @hameerabbasi.

**Comment by hameerabbasi** [comment]
Thanks @eric-wieser for guiding me through my first contribution to Numpy and literally hand-holding me on that other one. And thanks @mhvk for reviewing.

**Pull request merged**
